src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs

```diff
@@ -232,71 +226,262 @@ private static void Serialize(Stream output, FormatterEntry entry)
+            {
+                // Tags:
+                //   Tags count: 7-bit encoded int
+
+                var tagsCount = reader.Read7BitEncodedInt();
+                if (tagsCount > 0)
+                {
+                    //   For each tag:
+                    //     data byte length: 7-bit encoded int
+                    //     UTF-8 encoded byte[]
+                    for (var i = 0; i < tagsCount; i++)
+                    {
+                        reader.SkipString();
+                    }
+                }
+            }
+#endif
         }
-
-        // Body:
-        //   Segments count: 7-bit encoded int
-
-        var segmentsCount = reader.Read7BitEncodedInt();
-
-        //   For each segment:
-        //     data byte length: 7-bit encoded int
-        //     data byte[]
-
-        var segments = new List<byte[]>(segmentsCount);
-
-        for (var i = 0; i < segmentsCount; i++)
+        else
         {
-            var segmentLength = reader.Read7BitEncodedInt();
-            var segment = reader.ReadBytes(segmentLength);
+            // Body:
+            //   Bytes count: 7-bit encoded int
 
-            segments.Add(segment);
+            var payloadLength = checked((int)reader.Read7BitEncodedInt64());
+            if (payloadLength != 0)
+            {   // since the reader only supports linear memory currently, read the entire chunk as a single piece
+                result.SetBody(new(reader.ReadBytesMemory(payloadLength)), recycleBuffers: false); // we're reusing the live payload buffers
+            }
         }
 
-        result.Body = segments;
-
-        // Tags:
-        //   Tags count: 7-bit encoded int
-
-        var tagsCount = reader.Read7BitEncodedInt();
-
-        //   For each tag:
-        //     data byte length: 7-bit encoded int
-        //     UTF-8 encoded byte[]
-
-        var tags = new string[tagsCount];
+        Debug.Assert(reader.IsEOF, "should have read entire payload");
+        return result;
+    }
 
-        for (var i = 0; i < tagsCount; i++)
+    private static string ReadCommonHeader(ref FormatterBinaryReader reader)
+    {
+        int preamble = reader.Read7BitEncodedInt();
+        // LSB means "using common header/value"
+        if ((preamble & 1) == 1)
         {
-            var tagLength = reader.Read7BitEncodedInt();
-            var tagData = reader.ReadBytes(tagLength);
-            var tag = Encoding.UTF8.GetString(tagData);
-
-            tags[i] = tag;
+            // non-LSB is the index of the common header
+            return CommonHeaders[preamble >> 1];
         }
+        else
+        {
+            // non-LSB is the string length
+            return reader.ReadString(preamble >> 1);
+        }
+    }
 
-        result.Tags = tags;
-        return result;
+    static readonly string[] CommonHeaders = new string[]
+    {
+        // to remove values, use ""; DO NOT just remove the line, and do not re-order
```

Comment:
I think there needs to be a different story for how to remove a value.

If you removed one now (and added `""`) it would mean previously stored values would be read and end up with an empty header value, so we'd lose the value when returning a response.